cloud/smartos/smartos_image_source.py

```diff
@@ -0,0 +1,162 @@
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible. If not, see <http://www.gnu.org/licenses/>.
+#
+
+DOCUMENTATION = '''
+---
+module: smartos_image_source
+short_description: Manage image sources on SmartOS compute nodes.
+description:
+    - Add/delete image source on SmartOS compute nodes.
+version_added: "2.0"
+author: Adam Števko (@xen0l)
+options:
+    url:
+        description:
+            - URL of the image server.
+        required: true
+    type:
+        description:
+            - Specifies the type of the image server.
+        required: false
+        default: imgapi
+        choices: [ "imgapi", "docker", "dsapi" ]
+    state:
+        description:
+            - Set or reset the property value.
+        required: true
+        default: present
+        choices: [ "present", "absent" ]
+'''
+
+EXAMPLES = '''
+# Add datasets.at image source
+smartos_image_source: url=http://datasets.at state=present
+
+# Remove official image source
+smartos_image_source: url=https://images.joyent.com state=absent
+
+# Add docker registry as image source
+smartos_image_source: url=https://index.docker.io type=docker
+'''
+
+SUPPORTED_IMAGE_SOURCE_TYPES = ['imgapi', 'docker', 'dsapi']
+
+
+class ImageSource(object):
+
+    def __init__(self, module):
+        self.module = module
+
+        self.url = module.params['url']
+        self.type = module.params['type']
+        self.state = module.params['state']
+
+    def image_source_exists(self):
+        cmd = [self.module.get_bin_path('imgadm')]
+
+        cmd.append('sources')
+
+        if self.module.check_mode:
+            return True
+
+        (rc, out, _) = self.module.run_command(cmd)
+
+        if rc == 0 and self.url in out:
+            return True
+        else:
+            return False
+
+    def add_image_source(self):
+        cmd = [self.module.get_bin_path('imgadm')]
+
+        cmd.append('sources')
+        cmd.append('-a')
+        cmd.append(self.url)
+        cmd.append('-t')
+        cmd.append(self.type)
+
+        return self.module.run_command(cmd)
+
+    def delete_image_source(self):
+        cmd = [self.module.get_bin_path('imgadm')]
+
+        cmd.append('sources')
+        cmd.append('-d')
+        cmd.append(self.url)
+
+        return self.module.run_command(cmd)
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            url=dict(required=True),
+            type=dict(required=False, default='imgapi',
+                      choices=SUPPORTED_IMAGE_SOURCE_TYPES),
+            state=dict(
+                default='present', choices=['absent', 'present', 'reset']),
```

Comment:
So, what does "reset" do ? This is not handled by the code
( and if there is just present and absent, I think the condiction can later be simplified to just if/else rather than if/elsif)